File: lua/sum-of-multiples/sum-of-multiples.lua
local function sum_of_multiples(numbers)
    return {
        to = function(max)
            sum = 0
            for candidate = 1, max -1, 1 do
                for _, base in ipairs(numbers) do
                    if candidate % base == 0 then 
                        sum = sum + candidate
                        break
                    end 
                end
            end
            return sum
        end
    }
end


return sum_of_multiples

File: fsharp/sum-of-multiples/SumOfMultiples.fs
module SumOfMultiples

let multiplesTillOf' upperBound step = 
    step
    |> Seq.unfold (fun value ->
        if (value >= upperBound) then
            None
        else
            Some(value, value + step))
let sum' (numbers: int list) (upperBound: int): int = 
    numbers
    |> List.filter ((<) 0)
    |> List.map (multiplesTillOf' upperBound)
    |> List.fold Seq.append Seq.empty
    |> Seq.distinct 
    |> Seq.sum

let multiplesTillOf upperbound step =
    seq {step..step..upperbound-1}

let sum (numbers: int list) (upperBound: int): int = 
    numbers
    |> List.filter ((<) 0)
    |> Seq.collect (multiplesTillOf upperBound) // this is basically mapping and folding in one go
    |> Seq.distinct 
    |> Seq.sum